algorithms task (4)/student/include/impl.h

```diff
@@ -11,4 +13,7 @@
  * @return new slided range position
  */
 template <typename RandomAccessIterator>
-std::pair<RandomAccessIterator, RandomAccessIterator> slide(RandomAccessIterator begin, RandomAccessIterator last, RandomAccessIterator pos);
+std::pair<RandomAccessIterator, RandomAccessIterator> slide(RandomAccessIterator begin, RandomAccessIterator last, RandomAccessIterator pos)
+{
+    return std::make_pair(std::rotate(begin, last, pos), pos);
```

Comment:
_there can be more effective version with less amount of swaps :)_

Reply:
Hmmm...
As I understand it right the most effective std::rotate() implementation does N (or N-1, not so important) swaps, i.e. 3*N moves. Well, it can be done in much less moves, N + min(n1, n2), where n1, n2 are lengths of subranges, but it will require allocation of additional memory to hold temporary min(n1, n2) elements, it is almost the same I did in the first try. But we all know that allocating new memory is always heavy operation and most likely it will work slower than just N swaps in place.
What exactly did you mean by "there can be more effective version..." ??

Reply:
please check proposed solution (check your email)
just make this rotate conditionally more effective, don't move the whole range

Reply:
No, wait.

I gave up, I checked the solution. Now I see. But as I said, the task's statement (in "impl.h") is very confusing.
 * @param pos - new begin of slided range.
Here "pos" is new BEGIN of slided range in case of left shift, it is Ok and clear, BUT it is new END in case of right shift, and here is the reason of confusing. It would be clear if:
 * @param pos - new begin of slided range to the left, or new end of slided range to the right.
We all know that end is not begin... :)

More of that, unit tests, they test ONLY right shift case:
    if (last < pos) // right shift
    {
        return {std::rotate(begin, last, pos), pos};
    }
In case of left shift my solution will definitely crash (not acceptable). As well as everyone other's solution since everybody submitted the SAME solution (NO accepted solutions at all). What I want to say is that condition:
    if (pos < begin) // left shift
    {
        return {pos, std::rotate(pos, begin, last)};
    }
has never executed, because no unit tests for that case at all. And if it had tests, then nobody submitted current solution as it would crash and students would continue finding the right solution. It must be fixed.

See my changes in new update of the merge request. And with these new unit tests all solutions will crash (except your's of course).

Reply:
Oh, thanks, you're right. I will fix the tests, thanks!